include/xgboost/data.h

```diff
@@ -334,7 +325,7 @@ class BatchIterator {
   }
 
  private:
-  std::unique_ptr<BatchIteratorImpl<T>> impl_;
+  std::shared_ptr<BatchIteratorImpl<T>> impl_;
```

Comment:
Could you please point me to an example code where the old copy constructor is invoked?  I want to stay using unique_ptr if it's possible.

Reply:
it is right below in the batchset. any specific reason why you want to stay with `unique_ptr` ? there are so many allocations/deallocations in this code path...

Reply:
@sriramch I want to remove the strong reference between dmatrix and booster.  So users can free memory by deleting dmatrix object, from Python `del dtrain`.  See #4482 .  I'm not sure about how to do that, but shared_ptr seems to be a blocker.

Reply:
@trivialfis the scope of this change is restricted simply to the batchset that offers an iterator interface. as such, these objects should be transient in nature - meaning, they are created while someone is wanting to iterate over a dmatrix  and to get batches of some page type. the batch sets should soon be destructed after that. hence, i do not see how this can prolong the life of a dmatrix (as this isn't really encapsulating the state of the dmatrix, or even a part of it), even if someone is wanting to release it at some point (assuming that no one is releasing the dmatrix while iterating over it, which should be an invalid usage anyways).

Reply:
@sriramch Thanks for the explanation.  Looks good to me.